src/rpc/evo.cpp

```diff
@@ -1408,10 +1408,13 @@ static UniValue protx_list(const JSONRPCRequest& request, CDeterministicMNManage
             protx_list_help(request);
         }
 
-        LOCK(cs_main);
-
         bool detailed = !request.params[1].isNull() ? ParseBoolV(request.params[1], "detailed") : false;
 
+#ifdef ENABLE_WALLET
+        LOCK2(wallet ? wallet->cs_wallet : cs_main, cs_main);
```

Comment:
I don't understand... why is `cs_wallet` needed here exactly?

Reply:
> why is cs_wallet needed here exactly

There's call below `protx_list` -> `BuildDMNListEntry` -> `CheckWalletOwnsScript`:
```
return WITH_LOCK(pwallet->cs_wallet, return pwallet->IsMine(script)) == isminetype::ISMINE_SPENDABLE; 
```

and it can cause a deadlock due to wrong order of locks (cs_wallet supposed to be blocked in prior of cs_main)


> also, for wallet == nullptr this statement turns into LOCK2(cs_main, cs_main); which feels wrong
yes, that indeed feels wrong and ugly, but I haven't found other way to make it conditionally depends on `wallet` being null-ptr without significant code duplication

Reply:
> > why is cs_wallet needed here exactly
> 
> There's call below `protx_list` -> `BuildDMNListEntry` -> `CheckWalletOwnsScript`:
> 
> ```
> return WITH_LOCK(pwallet->cs_wallet, return pwallet->IsMine(script)) == isminetype::ISMINE_SPENDABLE; 
> ```
>  ...

I don't see this line anywhere https://github.com/dashpay/dash/blob/10869fff3b2cb7af5c4880c45e63bd85adec6084/src/rpc/evo.cpp#L1273-L1289

Reply:
oh, it's in #6003... hmm